test/CodeGenerator/AbstractInjectorTest.php

```diff
@@ -0,0 +1,197 @@
+
+        return new class($factoriesProvider, $injector, $container) extends AbstractInjector
+        {
+            private $provider;
+
+            public function __construct(
+                callable $provider,
+                InjectorInterface $injector,
+                ContainerInterface $container = null
+            ) {
+                $this->provider = $provider;
+                parent::__construct($injector, $container);
+            }
+
+            public function exposeFactoryValue(string $key)
+            {
+                return $this->factories[$key] ?? null;
+            }
+
+            protected function loadFactoryList() : void
+            {
+                $this->factories = ($this->provider)();
+            }
+        };
+    }
+
+    public function testImplementsContract()
+    {
+        $prophecy = $this->prophesize(InvokableInterface::class);
+        $prophecy->__call('__invoke', [])
+            ->shouldBeCalled()
+            ->willReturn([
+                'SomeService' => 'SomeFactory'
+            ]);
+
+        $subject = $this->createTestSubject($prophecy->reveal());
+        $this->assertInstanceOf(InjectorInterface::class, $subject);
+    }
+
+    public function testCanCreateReturnsTrueWhenAFactoryIsAvailable()
+    {
+        $className = uniqid('SomeClass');
+        $provider = function () use ($className) {
+            return [$className => 'SomeClassFactory'];
+        };
+
+        $this->decoratedInjectorProphecy
+            ->canCreate($className)
+            ->shouldNotBeCalled();
+
+        $subject = $this->createTestSubject($provider);
+        $this->assertTrue($subject->canCreate($className));
+    }
+
+    public function testCanCreateUsesDecoratedInjectorWithoutFactory()
+    {
+        $missingClass = uniqid('SomeClass');
+        $existingClass = uniqid('SomeOtherClass');
+        $provider = function () {
+            return [];
+        };
+
+        $this->decoratedInjectorProphecy
+            ->canCreate($missingClass)
+            ->shouldBeCalled()
+            ->willReturn(false);
+
+        $this->decoratedInjectorProphecy
+            ->canCreate($existingClass)
+            ->shouldBeCalled()
+            ->willReturn(true);
+
+        $subject = $this->createTestSubject($provider);
+
+        $this->assertTrue($subject->canCreate($existingClass));
+        $this->assertFalse($subject->canCreate($missingClass));
+    }
+
+    public function testCreateUsesFactory()
+    {
+        $factory = $this->prophesize(FactoryInterface::class);
+        $className = uniqid('SomeClass');
+        $params = ['someArg' => uniqid()];
+        $expected = new stdClass();
+        $provider = function () use ($className, $factory) {
+            return [$className => $factory->reveal()];
+        };
+
+        $factory->create(
+            $this->containerProphecy->reveal(),
+            $params
+        )
```

Comment:
Move the method call to the next line so it's more clear that the following lines are related:

```php
$factory
    ->create(
        $this->containerProphecy->reveal(),
        $params
    )
    ->shouldBeCalled()
    ->willReturn($expected);
```